### PR TITLE
Minor updates to `Pipeline`, `TilePipeline`, and `Tile`

### LINKED ISSUE #872: Tile pipeline state
When resetting a tile pipeline any attempt to get a sender with read() or readwrite() should fail.

A function getState() can be provided to the user to allow to check for the pipeline state.

### PR DESCRIPTION
**Pull request by msimberg**
- Add `reset` and `valid` member functions to `TilePipeline`
- Add `reset` and `valid` member functions to `Pipeline`
- Add default constructors to `Tile` and `TileData`
- Implement `RetiledMatrix::done` in terms of `TilePipeline::reset`

Note, I named it `{Tile,}Pipeline::reset` to mirror `optional::reset`. I'm indifferent about `RetiledMatrix::done` and happy to leave it as `done` since it's a higher level abstraction. I also did not add separate assertions in `RetiledMatrix::read{,write}` but rely on the assertions in `TilePipeline`, though it might make sense to add the assertions also to `RetiledMatrix`.

Fixes #872.

### CODE REVIEW

**Comment by msimberg** [comment]
cscs-ci run

File: include/dlaf/common/pipeline.h [hunk 1]
```diff
@@ -42,10 +43,25 @@ class Pipeline {
   ///
   /// @return a sender that will become ready as soon as the previous user releases the resource.
```
Comment (rasolca) [inline include/dlaf/common/pipeline.h:44]: Document the precondition (in other places as well)

File: include/dlaf/common/pipeline.h [hunk 2]
```diff
@@ -42,10 +43,25 @@ class Pipeline {
...
   Sender operator()() {
-    return pipeline.readwrite();
+    DLAF_ASSERT(pipeline.has_value(), "");
```
Comment (rasolca) [inline include/dlaf/common/pipeline.h:46]:
```suggestion
    DLAF_ASSERT(valid(), "");
```

File: include/dlaf/matrix/internal/tile_pipeline.h [hunk 3]
```diff
@@ -22,19 +23,57 @@ class TilePipeline {
...
-    return pipeline.read();
+    DLAF_ASSERT(valid(), "");
+    return pipeline->read();
   }
 
+
+  /// Get a read-write sender to a tile.
+  ///
+  /// The returned sender will send its value when the previous access to the
+  /// til has completed.
```
Comment (rasolca) [inline include/dlaf/matrix/internal/tile_pipeline.h]:
```suggestion
  /// tile has completed.
```
(multiple places)

File: include/dlaf/common/pipeline.h (hunk 1 above)
Comment (msimberg) [reply include/dlaf/common/pipeline.h:44]: https://github.com/eth-cscs/DLA-Future/pull/881/commits/c47688b2ff2394015fdea3c08259dc332e83e805

File: include/dlaf/common/pipeline.h (hunk 2 above)
Comment (msimberg) [reply include/dlaf/common/pipeline.h]: https://github.com/eth-cscs/DLA-Future/pull/881/commits/c47688b2ff2394015fdea3c08259dc332e83e805

File: include/dlaf/matrix/internal/tile_pipeline.h (hunk 3 above)
Comment (msimberg) [reply include/dlaf/matrix/internal/tile_pipeline.h]: https://github.com/eth-cscs/DLA-Future/pull/881/commits/6eb0c87e7ffb35f5b771745ef7c6f7ecdafaaf83

**Comment by msimberg** [comment]
cscs-ci run

**Review by rasolca — APPROVED** [review]

**Pull request merged**